packages/dds/tree/src/test/shared-tree/fuzzEditGenerator.ts

```diff
@@ -0,0 +1,254 @@
+        return {
+            fuzzType: "delete",
+            path,
+        };
+    }
+
+    async function setPayloadGenerator(state: EditState): Promise<FuzzSetPayload> {
+        const trees = state.testTreeProvider.trees;
+        const tree = trees[state.treeIndex];
+        // generate edit for that specific tree
+        const path = getExistingRandomNodePosition(tree, state.random);
+        return {
+            fuzzType: "setPayload",
+            path,
+            value: state.random.integer(Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER),
+        };
+    }
+
+    const baseEditGenerator = createWeightedAsyncGenerator<FuzzChange, EditState>([
+        [insertGenerator, 5],
+        [deleteGenerator, 1, ({ testTreeProvider, treeIndex }) => containsAtLeastOneNode(testTreeProvider.trees[treeIndex])],
+        [setPayloadGenerator, 1, ({ testTreeProvider, treeIndex }) => containsAtLeastOneNode(testTreeProvider.trees[treeIndex])],
+    ]);
+
+    return async (state: FuzzTestState): Promise<Operation | typeof done> => {
+        const trees = state.testTreeProvider.trees;
+        // does not include last tree, as we want a passive client
+        const treeIndex = state.random.integer(0, trees.length - 2);
+        const contents = await baseEditGenerator({
+            ...state,
+            treeIndex,
+        });
+        state.numberOfEdits += 1;
+        if (contents === done) {
+            return done;
+        }
+        return { type: "edit", contents, index: treeIndex };
+    };
+};
+
+export function makeOpGenerator(): AsyncGenerator<Operation, FuzzTestState> {
+    const opWeights: AsyncWeights<Operation, FuzzTestState> = [
+        [makeEditGenerator(), 3],
+        [{ type: "synchronize" }, 1],
+    ];
+    return createWeightedAsyncGenerator(opWeights);
+}
+
+const moves = {
+    field: ["enterNode", "nextField"],
+    nodes: ["stop", "firstField"],
+};
+
+function getRandomPlace(tree: ISharedTree, random: IRandom): UpPath {
+    const testerKey: FieldKey = brand("Test");
+    const cursor = tree.forest.allocateCursor();
+    moveToDetachedField(tree.forest, cursor);
+    const firstNode = cursor.firstNode();
+    if (!firstNode) {
+        // no node exists, insert a rootnode
+        cursor.free();
+        return { parent: undefined, parentField: rootFieldKeySymbol, parentIndex: 0 };
+    }
+    let currentPath = cursor.getPath();
+    assert(currentPath !== undefined)
+    const parentPath: UpPath = currentPath;
+    const firstField = cursor.firstField();
+    if (!firstField) {
+        // no fields, insert at random field at index 0 under rootnode
+        cursor.free();
+        return { parent: parentPath, parentField: testerKey, parentIndex: 0 };
+    }
+    currentPath = getExistingRandomNodePosition(tree, random);
+    const choosePath = random.pick(["currentNode", "nextNode", "addField"])
+    switch (choosePath) {
+        case "currentNode":
```

Comment:
Since we're picking a place as opposed to a node, it would clearer to name this case "beforeNode".